woocommerce/woocommerce
# [Experimental] [CYS - Full Composability]: Add tooltip when user clicks on pattern

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:





This PR adds a tooltip when the user clicks on a pattern.

Closes #47455 #47453.


This PR adds the tooltip when the user clicks on a pattern. This work required some refactoring, given that I had to optimize the functions that are called during each render to avoid a sluggish user experience.

### `usePopoverHandler` 

I added `usePopoverHandle` hook to handle the computation of the position of the popover based on the pointer

### `useAddAutoBlockPreviewEventListenersAndObservers` 

I added `useAddAutoBlockPreviewEventListenersAndObservers` to centralize all the hooks and mutationObservers in one place. Furthermore, I ensured that mutation observers are called only when necessary, given that they hurt performance (example: https://github.com/woocommerce/woocommerce/blob/693b153fc00ffbb9d9758f69c3e81c16c28c38a3/plugins/woocommerce-admin/client/customize-store/assembler-hub/hooks/auto-block-preview-event-listener.ts/#L258-L259).

Also, I ensured that we unsubscribe them when the component is unmounted. This is really important because the `AutoHeightBlockPreview` component is used for the main iframe and the small previews in the sidebar.

Futhermore, I created two new functions:

- `addInertToAllInnerBlocks`: this function adds the `inert`attribute to all the inner blocks, in this way, the browser will ignore those elements: https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/inert.
- `updateSelectedBlock`: updates the selected block when the user interacts with it.


### Removed not used code

Removed unused code related to the "other pages" feature, which wasn't shipped in the initial CYS version: https://github.com/woocommerce/woocommerce/pull/47583/files#diff-2320f0b44212379f7b24fd69a40a3b7ab6be829ea02c507d3c81de35c4948cc5L128-L165




### How to test the changes in this Pull Request:



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Ensure that you have the last version of Gutenberg enabled
2. Make sure you have the WooCommerce beta tester plugin installed
3. Head over to Tools > WCA Test Helper > Features
4. Make sure you see the `pattern-toolkit-full-composability` on the list.
5. Enable it.
6. Head over to WooCommerce > Home > Customize your store.
7. On the main iframe, click on a block.
8. Ensure that a tooltip appears.
9. Move the mouse pointer and ensure that the tooltip follows the pointer.
10. Hover another pattern.
11. Ensure that the tooltip disappears.
12. Click another pattern.
13. Ensure that the tooltip position changes.
14. Move the pointer on the left sidebar.
15. Ensure that the tooltip disappears.




https://github.com/woocommerce/woocommerce/assets/4463174/1fd3ded5-7257-4b3d-8945-9222cf86356f





### Changelog entry




<details>

#### Significance




#### Type




#### Message 

CYS: Show popover when the user clicks on the pattern

#### Comment 

</details>
